torchbenchmark/microbenchmarks/__init__.py

```diff
@@ -0,0 +1,120 @@
+    old_gpu_fuse = torch._C._jit_can_fuse_on_gpu()
+    old_texpr_fuser_state = torch._C._jit_texpr_fuser_enabled()
+    old_nvfuser_state = torch._C._jit_nvfuser_enabled()
+
+    torch._C._jit_override_can_fuse_on_cpu(False)
+    torch._C._jit_override_can_fuse_on_gpu(False)
+    torch._C._jit_set_texpr_fuser_enabled(False)
+    torch._C._jit_set_nvfuser_enabled(False)
+
+    try:
+        yield
+    finally:
+        torch._C._jit_override_can_fuse_on_cpu(old_cpu_fuse)
+        torch._C._jit_override_can_fuse_on_gpu(old_gpu_fuse)
+        torch._C._jit_set_texpr_fuser_enabled(old_texpr_fuser_state)
+        torch._C._jit_set_nvfuser_enabled(old_nvfuser_state)
+
+
+def make_tensor_from_type(inp_type: torch._C.TensorType):
+    if inp_type.requires_grad() is not False:
+        raise NotImplementedError("Tensors with requires_grad are not implemented")
+    return make_tensor(
+        inp_type.sizes(),
+        dtype=inp_type.dtype(),
+        device=inp_type.device())
+
+
+def load_graph_and_inputs(ir: str) -> Tuple[Any, List[Any]]:
+    graph = torch._C.parse_ir(ir)
+    graph.makeMultiOutputIntoTuple()
+    inputs = []
+    for inp in graph.inputs():
+        if isinstance(inp.type(), torch._C.FloatType):
+            inputs.append(random.uniform(.1, 100))
+        elif isinstance(inp.type(), torch._C.IntType):
+            inputs.append(random.randint(1, 100))
+        elif isinstance(inp.type(), torch._C.TensorType):
+            inputs.append(make_tensor_from_type(inp.type()))
+        else:
+            raise NotImplementedError(f"A default value is not implemented for type {inp.type()}")
+
+    func = torch._C._create_function_from_graph("forward", graph)
+    torch._C._jit_pass_erase_shape_information(func.graph)
+    return (func, inputs)
+
+
+def time_cuda(fn, inputs, test_runs):
+    start_event = torch.cuda.Event(enable_timing=True)
+    end_event = torch.cuda.Event(enable_timing=True)
+    torch.cuda.synchronize()
+    start_event.record()
+    torch.cuda.synchronize()
+    for i in range(test_runs):
+        fn(*inputs)
+        torch.cuda.synchronize()
+    end_event.record()
+    torch.cuda.synchronize()
+    return start_event.elapsed_time(end_event) / test_runs
+
+def time_cpu(fn, inputs, test_runs):
+    s = time.perf_counter()
+    for _ in range(test_runs):
+        fn(*inputs)
+    e = time.perf_counter()
+    return (e - s) / test_runs
+
+
+def run_test(ir, inputs, *, warmup_runs=10, test_runs=20) -> float:
+    graph, _ = load_graph_and_inputs(ir)
+    for _ in range(warmup_runs):
+        graph(*inputs)
+
+    is_cpu = None
+    for input in inputs:
+        if isinstance(input, torch.Tensor):
+            is_cpu = input.device.type == "cpu"
+            break
+    assert is_cpu != None
+
+    out = time_cpu(graph, inputs, test_runs) if is_cpu else time_cuda(graph, inputs, test_runs)
+    return out
+
+
+class NVFuserBenchmark():
+    def __init__(self, name, ir, warmup_runs = 10, test_runs = 20):
```

Comment:
It will be better to pass the `--nvfuser` argument as `extra_args` here, since `--nvfuser` is a benchmark-specific option

Reply:
@xuzhao9 to clarify - currently I'm just using `--fusers` to provide a list of all the fusers that should be attempted (that way you can compare the performance of whichever fusers you want to see; e.g. you might want to compare nvfuser to no fusions, or compare nvfuser to nnc). Is your suggestion to instead use flags like `--nvfuser` extra_arg to specify which specific fuser to use?

Reply:
> @xuzhao9 to clarify - currently I'm just using `--fusers` to provide a list of all the fusers that should be attempted (that way you can compare the performance of whichever fusers you want to see; e.g. you might want to compare nvfuser to no fusions, or compare nvfuser to nnc). Is your suggestion to instead use flags like `--nvfuser` extra_arg to specify which specific fuser to use?

Actually no, I am suggesting in the future, we may want to add more microbenchmarks besides this `nvfuser`, and the `nvfuser` option should be a "benchmark-specific" option and not be placed in `run.py`. You can use `parse_known_args()` in `run.py`, and deal with benchmark-specific options in `microbenchmarks/nvfuser/__init__.py`.